java-checks/src/main/java/org/sonar/java/checks/security/CryptographicKeySizeCheck.java

```diff
@@ -0,0 +1,133 @@
+ */
+package org.sonar.java.checks.security;
+
+import com.google.common.collect.ImmutableList;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.Optional;
+import org.sonar.check.Rule;
+import org.sonar.java.checks.methods.AbstractMethodDetection;
+import org.sonar.java.matcher.MethodMatcher;
+import org.sonar.java.model.LiteralUtils;
+import org.sonar.plugins.java.api.semantic.Symbol;
+import org.sonar.plugins.java.api.tree.AssignmentExpressionTree;
+import org.sonar.plugins.java.api.tree.ExpressionTree;
+import org.sonar.plugins.java.api.tree.IdentifierTree;
+import org.sonar.plugins.java.api.tree.LiteralTree;
+import org.sonar.plugins.java.api.tree.MemberSelectExpressionTree;
+import org.sonar.plugins.java.api.tree.MethodInvocationTree;
+import org.sonar.plugins.java.api.tree.Tree;
+import org.sonar.plugins.java.api.tree.VariableTree;
+
+@Rule(key = "S4426")
+public class CryptographicKeySizeCheck extends AbstractMethodDetection {
+
+  private static final String KEY_PAIR_GENERATOR = "java.security.KeyPairGenerator";
+  private static final String KEY_GENERATOR = "javax.crypto.KeyGenerator";
+  private static final String GET_INSTANCE_METHOD = "getInstance";
+  private static final String STRING = "java.lang.String";
+
+  private static final GetInstanceMethodInfo getInstanceInfo = new GetInstanceMethodInfo();
+
+  private static final Map<String, Integer> algorithmKeySizeMap = new HashMap<>();
+  static {
+    algorithmKeySizeMap.put("RSA", 2048);
+    algorithmKeySizeMap.put("Blowfish", 128);
+  }
+
+  @Override
+  protected List<MethodMatcher> getMethodInvocationMatchers() {
+    return ImmutableList.of(
+      MethodMatcher.create().typeDefinition(KEY_GENERATOR).name(GET_INSTANCE_METHOD).addParameter(STRING),
+      MethodMatcher.create().typeDefinition(KEY_PAIR_GENERATOR).name(GET_INSTANCE_METHOD).addParameter(STRING),
+      MethodMatcher.create().typeDefinition(KEY_PAIR_GENERATOR).name("initialize").withAnyParameters(),
+      MethodMatcher.create().typeDefinition(KEY_GENERATOR).name("init").withAnyParameters());
+  }
+
+  @Override
+  protected void onMethodInvocationFound(MethodInvocationTree mit) {
+    if (GET_INSTANCE_METHOD.equals(mit.symbol().name())) {
+      ExpressionTree argument = mit.arguments().get(0);
+      if (argument.is(Tree.Kind.STRING_LITERAL)) {
+        String algorithm = LiteralUtils.trimQuotes(((LiteralTree) argument).value());
+        if (algorithmKeySizeMap.containsKey(algorithm)) {
+          storeAlgorithmKeySize(algorithm);
+          storeSymbolOfAssignedInstance(mit);
+        }
+      }
+    } else {
+      if (isValidKeyGenSymbolOfInitializeMethod(mit)) {
+        checkAlgorithmParameterToReport(mit.arguments().get(0)).ifPresent(reportString -> reportIssue(mit, reportString));
+      }
+    }
+  }
+
+  private static boolean isValidKeyGenSymbolOfInitializeMethod(MethodInvocationTree mit) {
```

Comment:
I find the name a little confusing because `a.keyPairG.initialize(1024);` is called on a valid symbol from a Java perspective, but not from the rule perspective...

maybe `initializeIsCalledOnPreviouslyDefinedKeyGen` or `keyGenSymbolIsInClassScope` (or something without _valid_ in it)?